core/src/serde/de.rs

```diff
@@ -0,0 +1,693 @@
+            }
+        }
+
+        deserializer.deserialize_any(IpldVisitor)
+    }
+}
+
+macro_rules! impl_deserialize_integer {
+    ($ty:ident, $deserialize:ident, $visit:ident) => {
+        fn $deserialize<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+            match self {
+                Self::Integer(integer) => {
+                    if integer >= $ty::MIN.into() && integer <= $ty::MAX.into() {
+                        visitor.$visit(integer as $ty)
+                    } else {
+                        error(format!(
+                            "`Ipld::Integer` value was bigger than `{}`",
+                            stringify!($ty)
+                        ))
+                    }
+                }
+                _ => error(format!(
+                    "Only `Ipld::Integer` can be deserialized to `{}`, input was `{:#?}`",
+                    stringify!($ty),
+                    self
+                )),
+            }
+        }
+    };
+}
+
+/// A Deserializer for CIDs.
+///
+/// A separate deserializer is needed to make sure we always deserialize only CIDs as `Ipld::Link`
+/// and don't deserialize arbitrary bytes.
+struct CidDeserializer(Cid);
+
+impl<'de> de::Deserializer<'de> for CidDeserializer {
+    type Error = SerdeError;
+
+    #[inline]
+    fn deserialize_any<V: de::Visitor<'de>>(self, _visitor: V) -> Result<V::Value, Self::Error> {
+        unreachable!()
+    }
+
+    fn deserialize_bytes<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        visitor.visit_bytes(&self.0.to_bytes())
+    }
+
+    forward_to_deserialize_any! {
+        bool byte_buf char enum f32 f64  i8 i16 i32 i64 identifier ignored_any map newtype_struct
+        option seq str string struct tuple tuple_struct  u8 u16 u32 u64 unit unit_struct
+    }
+}
+
+/// Deserialize from an [`Ipld`] enum into a Rust type.
+///
+/// The deserialization will return an error if you try to deserialize into an integer type that
+/// would be too small to hold the value stored in [`Ipld::Integer`].
+///
+/// [`Ipld::Floats`] are cast, possibly with loss of precision, to the requested Rust type.
+impl<'de> de::Deserializer<'de> for Ipld {
+    type Error = SerdeError;
+
+    #[inline]
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        match self {
+            Self::Null => visitor.visit_none(),
+            Self::Bool(bool) => visitor.visit_bool(bool),
+            Self::Integer(i128) => visitor.visit_i128(i128),
+            Self::Float(f64) => visitor.visit_f64(f64),
+            Self::String(string) => visitor.visit_str(&string),
+            Self::Bytes(bytes) => visitor.visit_bytes(&bytes),
+            Self::List(list) => visit_seq(list, visitor),
+            Self::Map(map) => visit_map(map, visitor),
+            Self::Link(cid) => visitor.visit_newtype_struct(CidDeserializer(cid)),
```

Comment:
I've added a few more tests (the the most recent commit). `struct Wrapper(ByteBuf)`, does *not* deserialize a CID. If you still have concerns, please provide concrete tests (can be pseudo code), what I should test for.

Reply:
> Hm. Well, we need to support deserializing IPLD -> IPLD...

While I was looking into this, I was actually wondering about that. Do we really need that? Does a `deserialize_any` implementation make sense, or should we just remove it? Why would I want to deserialize `Ipld` into `Ipld`? I mean, I have an `Ipld` object already.

Reply:
Ah, it's because we're not calling "deserialize any".

Reply:
Ok, so, to trigger this, some code would need to:

1. Call `deserialize_any`.
2. Get called by `visit_newtype_struct`.
3. Call `deserialize_bytes`.

At the moment, I can only get step 3 to call back into `deserialize_any` (which panics). I'm not sure if anyone is turning around and calling `deserialize_bytes` in this case, but there's no reason they _can't_.

However, it would still be safer to expect a call to `deserialize_newtype_struct` so we can _check_ that we're deserializing into the correct thing. I.e.:

1. This function passes a deserializer that _only_ succeeds if called with `deserialize_newtype_struct` with the special CID string.
2. `visit_newtype_struct` on the `Ipld` type calls back into `deserialize_newtype_struct`.

We won't get infinite recursion because we'd be passing a custom deserializer.

Reply:
> 2. `visit_newtype_struct` on the `Ipld` type calls back into `deserialize_newtype_struct`.

This is not clear to me. To which `deserialize_newtype_struct` is it calling back, to the original one?

> We won't get infinite recursion because we'd be passing a custom deserializer.

I lost count, how many deserializers would we have? Changing the `CidDeserializer` to one that accepts `deserialize_newtype_struct` only, or is this yet another one.

When I was developing this, I always hit the problem that eventually I need to call `deserialize_bytes` as I haven't found any other way to pass on the bytes of the CID into a visitor, which then needs to construct the `Cid`. Though it can well be that there is a way.

Reply:
> To which deserialize_newtype_struct is it calling back, to the original one?

A new one, unfortunately.

---

But I don't think that works. `visit_newtype_struct` really shouldn't call `deserialize_newtype_struct`. It should always be the other way around. The problem we're running into here is that we're calling `visit_newtype_struct` from `deserialize_any` where we don't have any type information.